pkg/services/apiserver/builder/helper.go

```diff
@@ -61,7 +61,8 @@ func SetupConfig(
 	buildVersion string,
 	buildCommit string,
 	buildBranch string,
-	optionalMiddlewares ...web.Middleware,
+	optionalMiddlewares []web.Middleware,
+	optionalMiddlewaresK8sDependent []web.Middleware,
```

Comment:
Not sure if I follow correctly here, how could you do the selection with this function?

Reply:
Yeah, i don't see how either. @toddtreece I would love to know what you meant.

http.Handler doesn't have a way of addressing its inner handlers via an index. Also, the before and after are tricky. Seems that at present, after is at the top (last index), but the before is actually somewhere in the middle after a bunch of defaults filters are applied, but not the DefaultBuildHandlerChain.

Reply:
i might be misunderstanding what the use case is, but i was thinking of the normal k8s filter `http.Handler` wrapper approach:

```go
func additionalMiddleWareFunc(h http.Handler) http.Handler {
  return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    // do something before with w,r
    h.ServeHTTP(w, r) // call default handler chain 
    // do something after with w,r
  })
}
```


```go
// filters.WithRequester needs to be after the K8s chain because it depends on the K8s user in context
handler = filters.WithRequester(handler)
handler = genericapiserver.DefaultBuildHandlerChain(handler, c)

// additionalMiddleWareFunc can now decide to add operations before or after the passed in handler
handler = additionalMiddleWareFunc(handler)
```

Reply:
Okay, I'm on the same page now. I think this is way more complex than having dedicated hooks, as you need to know the underlying implementation to configure it correctly.

Reply:
I also don't think we could make it work for this use case, I don't see how we could add some middleware before it, as at this point the call chain is already at a point past `genericapiserver.DefaultBuildHandlerChain` and as @charandas pointed out there is no way to step backwards in the chain.

Reply:
> as you need to know the underlying implementation to configure it correctly

isn't the same thing true of passing in two sets of middlewares? it seems like you need to understand where the filters are applied either way. it might be simpler to be explicit and pass in `buildHandlerChainFunc func(apiHandler http.Handler, c *Config) http.Handler` as an optional arg, and override the default (or an OSS variant) if it's set.

```go
handler := genericapiserver.DefaultBuildHandlerChain(handler, c)

// use supplied handler chain
if buildHandlerChainFunc != nil {
  handler = buildHandlerChainFunc(handler, c)
}
```

Reply:
> at this point the call chain is already at a point past genericapiserver.DefaultBuildHandlerChain and as @charandas pointed out there is no way to step backwards in the chain.

sorry, i think i missed this message while writing my response. it sounds like i might have mixed up your use case with the tenant aware data source response limiting middleware that was discussed earlier this week. i thought you needed the response to be written first as well, but after reading the slack message again it sounds like this is for rate limiting and not response limiting.

if that's correct, then passing in a custom `buildHandlerChainFunc` as described in the previous comment seems like the easiest approach to maintain between enterprise and OSS because there's no ambiguity about what might be added in the chain. up to you though, feel free to stick with the current approach if that seems easier to understand.

Reply:
@JohnnyQQQQ I will push something to your PR based on @toddtreece's thoughts. I think he has a point and with this `additionalMiddlewares` optional argument that I added earlier, I added a bad precedent. Basically, passing in [BuildHandlerChainFunc](https://github.com/grafana/grafana/blob/main/pkg/services/apiserver/builder/helper.go#L94) from up top in `SetupConfig` will let us totally build out the handler chain in enterprise the way we want to.

It's better than having 2 sets of optional middlewares.